src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -245,11 +243,72 @@ void QueryToOperatorTransformer::Visit(parser::InsertStatement *op) {
     op->select->Accept(this);
     insert_expr->PushChild(output_expr_);
     output_expr_ = insert_expr;
+    return;
+  }
+  // column_objects represents the columns for the current table as defined in
+  // its schema
+  auto column_objects = target_table->GetColumnObjects();
+  // INSERT INTO table_name VALUES (val1, val2, ...), (val_a, val_b, ...), ...
+  if (op->columns.empty()) {
+    for (const auto &values : op->insert_values) {
+      if (values.size() > column_objects.size()) {
+        throw CatalogException(
+            "ERROR:  INSERT has more expressions than target columns");
+      } else if (values.size() < column_objects.size()) {
+        for (oid_t i = values.size(); i != column_objects.size(); ++i) {
+          // check whether null values are allowed in the rest of the columns
+          if (column_objects[i]->IsNotNull()) {
+            // TODO: Add check for default value's existence for the current column
+            throw CatalogException(StringUtil::Format("ERROR:  null value in column \"%s\" violates not-null constraint", column_objects[i]->GetColumnName().c_str()));
+          }
+        }
+      }
+    }
   } else {
-    auto insert_expr = std::make_shared<OperatorExpression>(
-        LogicalInsert::make(target_table, &op->columns, &op->insert_values));
-    output_expr_ = insert_expr;
+    // INSERT INTO table_name (col1, col2, ...) VALUES (val1, val2, ...), ...
+    auto num_columns = op->columns.size();
+    for (const auto &tuple : op->insert_values) {  // check size of each tuple
+      if (tuple.size() > num_columns) {
+        throw CatalogException(
+            "ERROR:  INSERT has more expressions than target columns");
+      } else if (tuple.size() < num_columns) {
+        throw CatalogException(
+            "ERROR:  INSERT has more target columns than expressions");
+      }
+    }
+
+    // vector below contains column ids of the columns mentioned in the insert statement
+    std::vector<oid_t> specified;
```

Comment:
I require this vector to store the ids of the specified columns. This is needed because we also need to check the `not null` constraint for the unspecified ones, just like we did in the implicit case.